File: src/main/java/org/noxet/noxetserver/commands/games/misc/Game.java
package org.noxet.noxetserver.commands.games.misc;

import net.md_5.bungee.api.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.messaging.ErrorMessage;
import org.noxet.noxetserver.messaging.Message;
import org.noxet.noxetserver.messaging.SuccessMessage;
import org.noxet.noxetserver.minigames.GameDefinition;
import org.noxet.noxetserver.minigames.MiniGameController;
import org.noxet.noxetserver.minigames.MiniGameManager;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class Game implements TabExecutor {
    @Override
    @SuppressWarnings("NullableProblems")
    public boolean onCommand(CommandSender commandSender, Command command, String s, String[] strings) {
        if(strings.length == 0) {
            new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Missing argument: what way to manage mini-games.").send(commandSender);
            return false;
        }

        if(strings[0].equalsIgnoreCase("play")) {
            if(!(commandSender instanceof Player)) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Only players can join games.").send(commandSender);
                return true;
            }

            Player player = (Player) commandSender;

            if(strings.length < 2) {
                new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Missing argument: what game to play.").send(player);
                return true;
            }

            GameDefinition gameToPlay = null;

            for(GameDefinition gameDefinition : GameDefinition.values())
                if(gameDefinition.getOptions().getId().equals(strings[1])) {
                    gameToPlay = gameDefinition;
                    break;
                }

            if(gameToPlay == null) {
                new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "That is not a mini-game.").send(player);
                return true;
            }

            if(MiniGameManager.isPlayerBusyInGame(player)) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "You are already in a game.").send(player);
                return true;
            }

            if(!MiniGameManager.joinGame(player, gameToPlay))
                Objects.requireNonNull(MiniGameManager.createNewGame(gameToPlay)).addPlayer(player);

            return true;
        } else if(strings[0].equalsIgnoreCase("join")) {
            if(!(commandSender instanceof Player)) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Only players can join games.").send(commandSender);
                return true;
            }

            Player player = (Player) commandSender;

            if(strings.length < 2) {
                new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Missing argument: ID of game to join.").send(player);
                return true;
            }

            MiniGameController game = MiniGameManager.getGameFromId(strings[1]);

            if(game == null) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "There is no game by such ID. It may have ended.").send(player);
                return true;
            }

            if(MiniGameManager.isPlayerBusyInGame(player)) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "You are already in a game.").send(player);
                return true;
            }

            game.addPlayer(player);

            return true;
        } else if(strings[0].equalsIgnoreCase("leave")) {
            if(!(commandSender instanceof Player)) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Only players can leave games.").send(commandSender);
                return true;
            }

            Player player = (Player) commandSender;

            MiniGameController game = MiniGameManager.findPlayersOrSpectatorsGame(player);

            if(game == null) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "You are not in a game.").send(player);
                return true;
            }

            game.disbandPlayer(player); // Try to remove both as player and spectator.

            return true;
        } else if(strings[0].equalsIgnoreCase("spectate")) {
            if(!(commandSender instanceof Player)) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Only players can spectate games.").send(commandSender);
                return true;
            }

            Player player = (Player) commandSender;

            if(strings.length < 2) {
                new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Missing argument: game ID to spectate.").send(player);
                return true;
            }

            MiniGameController game = MiniGameManager.getGameFromId(strings[1]);

            if(game == null) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Invalid game ID. Is it already over?").send(player);
                return true;
            }

            if(MiniGameManager.isPlayerBusyInGame(player)) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "You are already in a game.").addButton("Leave", ChatColor.RED, "Leave this game", "game leave").send(player);
                return true;
            }

            game.addSpectator(player);

            return true;
        } else if(strings[0].equalsIgnoreCase("info")) {
            if(!(commandSender instanceof Player) && strings.length < 2) {
                new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Non-players must provide an argument for the game ID to see info about.").send(commandSender);
                return true;
            }

            MiniGameController game = strings.length > 1 ? MiniGameManager.getGameFromId(strings[1]) : MiniGameManager.findPlayersOrSpectatorsGame((Player) commandSender);

            if(game == null) {
                if(strings.length > 1) {
                    new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Invalid game ID.").send(commandSender);
                    return true;
                }

                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "You are not in a game. To see info about a game, you must either be in the game or provide the game ID as an argument.").send(commandSender);
                return true;
            }

            new Message(
                    "§3Game ID: §b" + game.getGameId() + "\n" +
                    "§3Game: §b" + game.getGame().getOptions().getDisplayName() + "\n" +
                    "§3Status: §b" + game.getState() + "\n" +
                    "§3Players: §b" + game.getPlayers().size() + " / " + game.getGame().getOptions().getMaxPlayers() + "\n" +
                    "§3Spectators: §b" + game.getSpectators().size() + "\n"
            ).addButton(
                    "Join",
                    ChatColor.GREEN,
                    "Join this game",
                    "game join " + game.getGameId()
            ).addButton(
                    "Spectate",
                    ChatColor.GOLD,
                    "Spectate this game",
                    "game spectate " + game.getGame()
            ).addButton(
                    "Stop",
                    ChatColor.RED,
                    "Soft stop this game",
                    "game stop soft " + game.getGameId()
            ).addButton(
                    "Hard-stop",
                    ChatColor.DARK_RED,
                    "Hard stop this game (instant stop)",
                    "game stop hard " + game.getGameId()
            ).send(commandSender);

            return true;
        } else if(strings[0].equalsIgnoreCase("list")) {
            new Message("§3Running games: " + MiniGameManager.getRegisteredGames().size()).send(commandSender);

            for(MiniGameController game : MiniGameManager.getRegisteredGames())
                new Message(
                        "└§3§l" + game.getGame().getOptions().getDisplayName() + " §b" + game.getGameId()
                ).addButton(
                        "Info",
                        ChatColor.YELLOW,
                        "Read more about this game",
                        "game info " + game.getGameId()
                ).send(commandSender);

            return true;
        } else if(strings[0].equalsIgnoreCase("stop")) {
            if(!commandSender.isOp()) {
                new ErrorMessage(ErrorMessage.ErrorType.PERMISSION, "Only operators can stop games.").send(commandSender);
                return true;
            }

            if(strings.length < 2) {
                new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Missing argument: stop mode (soft/hard).").send(commandSender);
                return true;
            }

            boolean hard;

            switch(strings[1].toLowerCase()) {
                case "hard":
                    hard = true;
                    break;
                case "soft":
                    hard = false;
                    break;
                default:
                    new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Invalid argument: stop mode (soft/hard).").send(commandSender);
                    return true;
            }

            if(strings.length < 3) {
                new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Missing argument: game ID to stop.").send(commandSender);
                return true;
            }

            if(strings[2].equals("*")) {
                new Message("Stopping all games " + (hard ? "hard" : "soft") + "... (" + MiniGameManager.getRegisteredGames().size() + ")").send(commandSender);

                int maxTicks = 0;

                for(MiniGameController game : MiniGameManager.getRegisteredGames())
                    if(hard)
                        game.stop();
                    else
                        maxTicks = Math.max(game.softStop(), maxTicks);

                if(!hard)
                    new Message("§4All games will have been stopped in " + maxTicks / 20 + " seconds.").send(commandSender);

                new BukkitRunnable() {
                    @Override
                    public void run() {
                        new SuccessMessage("All games have been stopped.").send(commandSender);
                    }
                }.runTaskLater(NoxetServer.getPlugin(), maxTicks);

                return true;
            }

            MiniGameController game = MiniGameManager.getGameFromId(strings[2]);

            if(game == null) {
                new ErrorMessage(ErrorMessage.ErrorType.COMMON, "Invalid game ID.").send(commandSender);
                return true;
            }

            new Message("§cStopping game " + (hard ? "hard (instantly)" : "soft (waiting for handler)") + "...").send(commandSender);

            int ticks;

            if(hard) {
                ticks = 0;
                game.stop();
            } else {
                ticks = game.softStop();
                new Message("§4Soft stop margin: " + ticks / 20 + " seconds.").send(commandSender);
            }

            new BukkitRunnable() {
                @Override
                public void run() {
                    new SuccessMessage("Game " + game.getGameId() + " stopped.").send(commandSender);
                }
            }.runTaskLater(NoxetServer.getPlugin(), ticks);

            return true;
        } else {
            new ErrorMessage(ErrorMessage.ErrorType.ARGUMENT, "Invalid subcommand: '" + strings[0] + "'.").send(commandSender);
            return false;
        }
    }

    @Override
    @SuppressWarnings("NullableProblems")
    public List<String> onTabComplete(CommandSender commandSender, Command command, String s, String[] strings) {
        List<String> completions = new ArrayList<>();

        if(strings.length == 1) {
            completions.addAll(Arrays.asList("play", "join", "leave", "spectate", "info", "list", "stop"));
        } else if(strings.length == 2) {
            switch(strings[0].toLowerCase()) {
                case "play":
                    for(GameDefinition eachGameDefinition : GameDefinition.values())
                        completions.add(eachGameDefinition.getOptions().getId());
                    break;
                case "stop":
                    completions.addAll(Arrays.asList("soft", "hard"));
                    break;
            }
        } else if(strings.length == 3) {
            if(strings[0].equalsIgnoreCase("stop"))
                completions.add("*");
        }

        return completions;
    }
}

File: src/main/java/org/noxet/noxetserver/commands/games/misc/Game.scala
package org.noxet.noxetserver.commands.games.misc

import org.bukkit.command.{Command, CommandSender, TabExecutor}
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitRunnable
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.commands.RegisteredCommand
import org.noxet.noxetserver.messaging.ErrorMessage.ErrorType.*
import org.noxet.noxetserver.messaging.{ErrorMessage, Message, SuccessMessage}
import org.noxet.noxetserver.minigames.party.Party
import org.noxet.noxetserver.minigames.worldeater.WorldEater
import org.noxet.noxetserver.minigames.{GameDefinition, MiniGameController, MiniGameManager}

object Game extends TabExecutor, RegisteredCommand("game", this):
  override def onCommand(commandSender: CommandSender, command: Command, s: String, strings: Array[String]): Boolean =
    strings match
      case Array() =>
        ErrorMessage(Argument, "Missing argument: what way to manage mini-games.").send(commandSender)
        false
      case Array("play") =>
        ErrorMessage(Argument, "Missing argument: what game to play.").send(player)
        true
      case Array("play", gameName) =>
        commandSender match
          case p: Player if MiniGameManager.isPlayerBusyInGame(p) =>
            ErrorMessage(Common, "You are already in a game.").send(player)
          case p: Player =>
            GameDefinition.values.find(_.options.getId == gameName) match
              case Some(gameToPlay) => MiniGameManager.playGame(p, gameToPlay)
              case None => ErrorMessage(Argument, s"No mini-game by such name: '$gameName'.").send(p)
          case _ =>
            ErrorMessage(Common, "Only players can join games.").send(commandSender)
        true
      case Array("join") =>
        ErrorMessage(Argument, "Missing argument: ID of game to join.").send(player)
        true
      case Array("join", gameId) =>
        commandSender match
          case p: Player if Party.isPlayerMemberOfParty(p) =>
            ErrorMessage(Common, "You are in a party, and cannot join games alone.").send(p)
          case p: Player if MiniGameManager.isPlayerBusyInGame(p) =>
            ErrorMessage(Common, "You are already in a game.").send(player)
          case p: Player =>
            MiniGameManager.getGameFromId(gameId) match
              case Some(game) => game.addPlayer(p)
              case None => ErrorMessage(Common, "No game with such ID. It may have ended.").send(p)
          case _ =>
            ErrorMessage(Common, "Only players can join games.").send(commandSender)
        true
      case Array("leave") =>
        commandSender match
          case p: Player =>
            MiniGameManager.findPlayersOrSpectatorsGame(player) match
              case Some(game) =>
                game.disbandPlayer(p) // Try to remove both as player and spectator.
                Message("§cYou left the game.").send(p)
              case None =>
                ErrorMessage(Common, "You are not in a game.").send(p)
          case _ =>
            ErrorMessage(Common, "Only players can leave games.").send(commandSender)
        true
      case Array("spectate") =>
        ErrorMessage(Argument, "Missing argument: game ID to spectate.").send(player)
        true
      case Array("spectate", gameId) =>
        commmandSender match
          case p: Player if MiniGameManager.isPlayerBusyInGame(p) =>
            ErrorMessage(Common, "You are already in a game.")
              .addButton(
                "Leave",
                ChatColor.RED,
                "Leave this game",
                "game leave")
              .send(p)
          case p: Player =>
            MiniGameManager.getGameFromId(gameId) match
              case Some(game) => game.addSpectator(player)
              case None =>
                ErrorMessage(Common, "Invalid game ID. Is it already over?").send(p)
          case _ =>
            ErrorMessage(Common, "Only players can spectate games.").send(commandSender)
        true
      case Array("info") =>
        commandSender match
          case p: Player if MiniGameManager.findPlayersOrSpectatorsGame(p).isDefined =>
            val game = MiniGameManager.findPlayersOrSpectatorsGame(p).get
            p.performCommand(s"game info ${game.gameId}")
          case _ =>
            ErrorMessage(Argument,
              "Unless you are in a game you must provide an argument " +
                "for the game ID to see info about.").send(commandSender)
        true
      case Array("info", gameId) =>
        MiniGameManager.getGameFromId(gameId) match
          case Some(game) =>
            Message(
              Map(
                "Game ID" -> game.gameId,
                "Game" -> game.game.options.getDisplayName,
                "Status" -> game.state,
                "Players" -> game.getPlayers.size + " / " + game.game.options.getMaxPlayers,
                "Spectators" -> game.getSpectators.size,
              ).map((k, v) => s"§3$k: §b$v")
            ).addButton(
              "Join",
              ChatColor.GREEN,
              "Join this game",
              s"game join ${game.gameId}"
            ).addButton(
              "Spectate",
              ChatColor.GOLD,
              "Spectate this game",
              s"game spectate ${game.gameId}"
            ).addButton(
              "Stop",
              ChatColor.RED,
              "Soft stop this game",
              s"game stop soft ${game.gameId}"
            ).addButton(
              "Hard-stop",
              ChatColor.DARK_RED,
              "Hard stop this game (instant stop)",
              s"game stop hard ${game.gameId}"
            ).send(commandSender)
          case None =>
            ErrorMessage(Common, "Invalid game ID.").send(commandSender)
        true
      case Array("list") =>
        Message(s"§3Running games: ${MiniGameManager.getRegisteredGames.size}").send(commandSender)

        for game <- MiniGameManager.getRegisteredGames do
          Message(
            s"└§3§l${game.game.options.getDisplayName} §b${game.gameId}"
          ).addButton(
            "Info",
            ChatColor.YELLOW,
            "Read more about this game",
            s"game info ${game.gameId}"
          ).send(commandSender)

        true
      case Array("stop", _*) if !commandSender.isOp =>
        ErrorMessage(Permission, "Only operators can stop games.").send(commandSender)
        true
      case Array("stop") =>
        ErrorMessage(Argument, "Missing argument: stop mode (soft/hard).").send(commandSender)
        true
      case Array("stop", mode) =>
        commandSender match
          case p: Player if MiniGameManager.findPlayersOrSpectatorsGame(p) =>
            val game = MiniGameManager.findPlayersOrSpectatorsGame(p).get
            p.performCommand(s"game stop ${game.gameId}")
          case _ =>
            ErrorMessage(Common, "Missing argument: game ID to stop. " +
              "In-game players/spectators can omit this argument.").send(commandSender)
        true
      case Array("stop", mode, "*") =>
        MiniGameManager.getRegisteredGames.foreach(g =>
          commandSender.performCommand(s"game stop $mode ${g.gameId}"))
      case Array("stop", mode, gameId) =>
        hard = mode match
          case "hard" => true
          case "soft" => false
          case _ =>
            ErrorMessage(Argument, "Invalid argument: stop mode (soft/hard).")
              .send(commandSender)
            return true

        MiniGameManager.getGameFromId(gameId) match
          case Some(game) =>
            Message("§cStopping game " +
              (if hard then "hard (instantly)"
              else "soft (waiting for handler)") +
              "...")
              .send(commandSender)

            val ticks =
              if hard then
                game.stop()
                0
              else
                game.softStop

            if ticks != 0 then
              Message(s"§4Soft stop margin: ${ticks / 20} seconds.").send(commandSender)

            game.scheduleTask(() =>
              SuccessMessage(s"Game ${game.gameId} stopped.").send(commandSender),
              Math.max(ticks - 1, 0))
          case None =>
            ErrorMessage(Common, "Invalid game ID.").send(commandSender)
        true
      case Array("debug-create") =>
        ErrorMessage(Argument, "Missing argument: what game to play.").send(player)
        true
      case Array("debug-create", gameName) =>
        commandSender match
          case s if !s.isOp =>
            ErrorMessage(Permission, "Only operators can create games.").send(commandSender)
          case p: Player if MiniGameManager.isPlayerBusyInGame(player) =>
            ErrorMessage(Common, "You are already in a game.").send(player)
          case p: Player =>
            GameDefinition.values.find(_.options.getId == gameName) match
              case Some(gameDef) =>
                val game = MiniGameManager.createNewGame(gameToPlay)
                game.addPlayer(player)
                // start() does not check for minimum player requirement, so the game will start anyway.
                game.start()
              case None =>
                ErrorMessage(Argument, "That is not a mini-game.").send(player)
          case _ =>
            ErrorMessage(Common, "Only players can create games.").send(commandSender)
        true
      case Array(subCommand, _*) =>
        ErrorMessage(Argument, s"Invalid subcommand: '${subCommand}'.")
          .send(commandSender)
        false

  override def onTabComplete(commandSender: CommandSender, command: Command, s: String, strings: Array[String]): List[String] =
    strings match
      case Array(_) =>
        List("play", "join", "leave", "spectate", "info", "list", "stop", "debug-create")
      case Array(subCommand, _) =>
        subCommand match
          case "play" | "debug-create" =>
            GameDefinition.values.map(_.options.getId)
          case "stop" =>
            List("soft", "hard")
      case Array(subCommand, _, _) =>
        subCommand match
          case "stop" => List("*")
          case _ => List()